### PR TITLE
[TS] LPP-47305 > LPS-169907 Fix the asset publisher can retrieve the URL parameter segment

### PR DESCRIPTION
**Pull request by ces-totrinh**
Hi guys,
Please help review this PR and leave your comments. Thanks.
Ticket: https://issues.liferay.com/browse/LPS-169907
cc: @hongvo2308

### CODE REVIEW

**Comment by tototrinh** [comment]
ci:test:sf

**Comment by tototrinh** [comment]
ci:test:relevant

**Comment by claraizquierdo** [comment]
ci:reevaluate:1357401_3149

**Comment by claraizquierdo** [comment]
ci:test:tango

**Comment by claraizquierdo** [comment]
ci:test:relevant

**Comment by claraizquierdo** [comment]
@CarlosBrichete could you have a look at the tests failing in the tango suit please?

**Comment by CarlosBrichete** [comment]
The test failing in CI are also failing in our daily routine, so I'd say they are not related to this PR :)

**Review by claraizquierdo — APPROVED** [review]
The code seems to work perfectly

**Review by Mikellorza — APPROVED** [review]

**Comment by Mikellorza** [comment]
ci:forward

File: modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/display/context/AssetPublisherDisplayContext.java [hunk 1]
```diff
@@ -2266,10 +2267,18 @@ private String _getSegmentsAnonymousUserId() {
...
-			_themeDisplay.getScopeGroupId(), _themeDisplay.getUserId(),
-			_requestContextMapper.map(
-				PortalUtil.getHttpServletRequest(_portletRequest)));
+		long[] requestSegmentsEntryIds = GetterUtil.getLongValues(
+			_portletRequest.getAttribute(SegmentsWebKeys.SEGMENTS_ENTRY_IDS));
+		long[] portletSegmentsEntryIds =
+			_segmentsEntryRetriever.getSegmentsEntryIds(
+				_themeDisplay.getScopeGroupId(), _themeDisplay.getUserId(),
+				_requestContextMapper.map(
+					PortalUtil.getHttpServletRequest(_portletRequest)));
```
Comment (lfbesada) [inline modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/display/context/AssetPublisherDisplayContext.java:2276]: @tototrinh @hongvo2308 Thanks for the fix! ❤️ 
LGTM, I have seen only a minor issue, could you please change this call to not use the deprecated method?
Thank you very much!

**Review by lfbesada — CHANGES_REQUESTED** [review]

**Comment by hongvo2308** [comment]
Hi @lfbesada 
I have updated this PR to use another method (which is not deprecated). Could you please double-check?
Thank you.

**Review by lfbesada — APPROVED** [review]

**Comment by claraizquierdo** [comment]
ci:forward

**Comment by hongvo2308** [comment]
Hi @claraizquierdo 
Since the CSE has asked about this PR, may I ask for the next steps for this PR? What should be the next steps for this one? And please let me know if it needs any updates from my end.
Thank you very much.

**Comment by claraizquierdo** [comment]
ci:forward

**Comment by claraizquierdo** [comment]
Hi @hongvo2308 ,

looks like the CI has some flaky test and is not allowing us to forward the pull. @Luizcarlosqueiroz could you confirm that the tests failing are not related to this pull. @lfbesada , maybe QA from Echo can confirm too. @hongvo2308, If Tango QA team and Echo QA team confirm that everything is working correctly in this pull, you could send this PR to Brian's repository directly.

**Comment by lfbesada** [comment]
Hi @claraizquierdo 
I've looked at the test that fails and it doesn't seem to be related to me.
Regards

**Comment by Luizcarlosqueiroz** [comment]
Hey @claraizquierdo!
The tests from Tango are not related with this PR :smile:

**Comment by cgoncas** [comment]
ci:test:relevant

**Review by cgoncas — CHANGES_REQUESTED** [review]
Hey @tototrinh and @hongvo2308 

I have reviewed this fix and I don't fully understand the fix, because now the SEGMENTS_ENTRY_IDS  are now filled with all the segments of the user, also, debuguing it, the problem is that the parameters are not filled in the context when the segments are evaluated.

I have created another PR with my proposal, could you take a look and tell my what do you think 

https://github.com/liferay-tango/liferay-portal/pull/3119.

Thanks!!

File: modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/display/context/AssetPublisherDisplayContext.java [hunk 2]
```diff
@@ -2266,10 +2267,20 @@ private String _getSegmentsAnonymousUserId() {
 	}
 
 	private long[] _getSegmentsEntryIds() {
-		return _segmentsEntryRetriever.getSegmentsEntryIds(
-			_themeDisplay.getScopeGroupId(), _themeDisplay.getUserId(),
-			_requestContextMapper.map(
-				PortalUtil.getHttpServletRequest(_portletRequest)));
+		long[] requestSegmentsEntryIds = GetterUtil.getLongValues(
+			_portletRequest.getAttribute(SegmentsWebKeys.SEGMENTS_ENTRY_IDS));
```
[the remark this answers is not in the document]
Comment (hongvo2308) [reply modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/display/context/AssetPublisherDisplayContext.java:2271]: Hi @cgoncas 
I found that this attribute is filled at this line https://github.com/liferay/liferay-portal/blob/8618a3437b6fca76e6f909e8aad4fe0b17da0649/modules/apps/segments/segments-service/src/main/java/com/liferay/segments/internal/events/SegmentsServicePreAction.java#L115 
And the function is called with some conditions at https://github.com/liferay/liferay-portal/blob/8618a3437b6fca76e6f909e8aad4fe0b17da0649/modules/apps/segments/segments-service/src/main/java/com/liferay/segments/internal/events/SegmentsServicePreAction.java#L149 
So I think that there must be some restrict to use this.
Could you please double-check?
Thank you.

[the remark this answers is not in the document]
Comment (cgoncas) [reply modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/display/context/AssetPublisherDisplayContext.java:2271]: Yeahh @hongvo2308!!
 In the service pre action we only filled the property with the segments of the experience.

so this is something that can not be used.

[the remark this answers is not in the document]
Comment (hongvo2308) [reply modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/display/context/AssetPublisherDisplayContext.java:2271]: Hi @cgoncas 
I am not sure about the term "the segments of the experience". I guess it should be specific segments, is it about the Segment for the Blank/Content Page? Or it's something else and where it is used?
Thank you.

[the remark this answers is not in the document]
Comment (cgoncas) [reply modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/display/context/AssetPublisherDisplayContext.java:2271]: in the past, this property contains all the segments of the user, but it implies a huge performance problem, that is the reason you should obtain the segments of the user when it is needed.

The problem of the bug, is that the context of the segment is not correctly filled, so you need to fix this. Your solution is not a valid one, sorry :(.

[the remark this answers is not in the document]
Comment (cgoncas) [reply modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/display/context/AssetPublisherDisplayContext.java:2271]: As you can see in https://github.com/liferay/liferay-portal/blob/8618a3437b6fca76e6f909e8aad4fe0b17da0649/modules/apps/segments/segments-service/src/main/java/com/liferay/segments/internal/events/SegmentsServicePreAction.java#L107, we only obtain the segments that match the user between segmentsExperienceIdsSegmentsEntryIds, not all the segments of the user

[the remark this answers is not in the document]
Comment (hongvo2308) [reply modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/display/context/AssetPublisherDisplayContext.java:2271]: @cgoncas 
Thank you for the explanation. Now I have a better understanding of these features.
Thank you very much.

**Comment by hongvo2308** [comment]
Hi @cgoncas 
Thank you for your help and great explanation. My solution is not correct so feel free to close this PR.
Thank you very much.

**Comment by cgoncas** [comment]
> Hi @cgoncas Thank you for your help and great explanation. My solution is not correct so feel free to close this PR. Thank you very much.

I am closing this PR, and we will try to merge: https://github.com/liferay-tango/liferay-portal/pull/3119

**Pull request closed without merging**